### PR TITLE
Add Modal beam example

### PR DESCRIPTION
**Pull request by maxnezdyur**
Part of #27295

### CODE REVIEW

**Comment by maxnezdyur** [comment]
@lynnmunday 
Can you look at this.

**Comment by maxnezdyur** [comment]
Will have to wait on https://github.com/idaholab/moose/pull/27323.

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 1]
```diff
@@ -1,5 +1,93 @@
 # Frequency Domain Analysis
 
+For frequency domain analysis for Solid Mechanics MOOSE has two main methods
+modal analysis and frequency response functions. The following example will go
```
Comment (lynnmunday) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:4]:
```suggestion
The MOOSE Solid Mechanics module performs frequency domain analysis using modal analysis and frequency response functions. The following example will go
```

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 2]
```diff
@@ -1,5 +1,93 @@
...
+\begin{equation}
+(\mathbf{A} - \omega^2 \mathbf{B})\mathbf{u} = 0
+\end{equation}
+where $\mathbf{A}$ holds the "noneigen" kernels, $\mathbf{B}$ holds the "eigen" kernels, $\omega$ are the natural frequencies, and $\mathbf{u}$ are the mode shapes.
+In MOOSE, modal analysis is set up using the Eigenvalue Executioner. The
+Eigenvalue System allows you to define which kernels go into which matrix. It's important
+to note that in modal analysis, Neumann boundary conditions are
+ignored since we are interested in the free vibration response of the structure.
+Only Dirichlet boundary conditions are considered.
+
```
Comment (lynnmunday) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:28]: Will you move lines 136-153 here that describe the analytic solution.  Then your results will follow it.  The analytic results are $\omega_{1a}=$509Hz\\
$\omega_{1b}=$763Hz. Do you know if you are using the same coarse mesh as the coarse mesh FRF results below.  If so, natural frequencies are much better, you should mention that.

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 3]
```diff
@@ -1,5 +1,93 @@
...
+represent the mass terms and contribute to the "B" matrix. The
+[!param](/Kernels/CoefReaction/coefficient) in [kernels/CoefReaction.md] is set
+to a negative value which corresponds to a positive density. The
+`extra_vector_tags = 'eigen'` parameter is used to indicate that these kernels
+contribute to "B" matrix.
+
+!listing test/tests/modal_analysis/modal.i block=Kernels id=m_kernel caption=
+Kernels for A and B
+
+The BCs block defines the boundary conditions for the problem. In this example, we have Dirichlet boundary conditions applied to the left boundary for all displacement components. The [DirichletBC.md] is used for the "A" matrix, while the [EigenDirichletBC.md] is used for the "B" matrix.
```
Comment (lynnmunday) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:53]: Would the boundary conditions on the B matrix ever be different from the A matrix?  If not, mention something about that.  It seems like there should have a eigen tag for BCs too instead of a separate BC.

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 4]
```diff
@@ -1,5 +1,93 @@
...
+[!param](/Executioner/Eigenvalue/n_eigen_pairs) parameter sets the number of eigenvalue pairs to compute.
+
+!listing test/tests/modal_analysis/modal.i block=Executioner id=m_exec
+caption=Eigenvalue Executioner
+
+To output all the eigenvalues solved in the system we can use the
+[vectorpostprocessors/Eigenvalues.md] vectorpostprocessor. While we have solved for the two smallest
+eigenvalues, currently MOOSE only has the ability to output a single one. To
+change which eigenvector is outputted adjust the index in
+[!param](/Problem/EigenProblem/active_eigen_index).
```
Comment (lynnmunday) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:74]:
```suggestion
[vectorpostprocessors/Eigenvalues.md] vectorpostprocessor. While we have solved for the two smallest
eigenvalues, currently MOOSE only has the ability to output a single eigenvector. To
change which eigenvector is outputted adjust the index in
[!param](/Problem/EigenProblem/active_eigen_index) and rerun the simulation.
```

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 5]
```diff
@@ -1,5 +1,93 @@
...
+
+!listing test/tests/modal_analysis/modal.i block=VectorPostprocessors Problem
+id=m_prob caption=Eigenvalues
+
+For this coarse mesh example, we were able to determine the first two
+$\omega$'s: 552.8 Hz and 789 Hz. Which are close to the theoretical result of
+600 Hz and 800 Hz, and as the mesh is refined the results would converge on to the
```
Comment (lynnmunday) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:81]:
```suggestion
509 Hz and 763 Hz, and as the mesh is refined the results would converge on to the
```

Comment (lynnmunday) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:81]: Say that this is closer than the natural frequencies determined using the same mesh with and the frequency response function in the next section.

File: modules/solid_mechanics/test/tests/modal_analysis/modal.i [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  n_eigen_pairs = 2
+  n_basis_vectors = 5
+  petsc_options = '-eps_monitor_all -eps_view'
+  petsc_options_iname = '-st_type -eps_target -st_pc_type -st_pc_factor_mat_solver_type'
+  petsc_options_value = 'sinvert 0 lu mumps'
+  eigen_tol = 1e-8
+[]
+
+[VectorPostprocessors]
+  [omega_squared]
```
Comment (lynnmunday) [inline modules/solid_mechanics/test/tests/modal_analysis/modal.i:131]: Does it output the square of the natural frequency?

File: modules/solid_mechanics/test/tests/modal_analysis/modal.i [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    type = Eigenvalues
+    execute_on = TIMESTEP_END
+  []
+
+[]
+
+[Problem]
+  type = EigenProblem
+  active_eigen_index = ${index}
```
Comment (lynnmunday) [inline modules/solid_mechanics/test/tests/modal_analysis/modal.i:140]: This seems like a weird place to specify which eigenvector you want output in the exodus file.   Can you output the eigenvector displacement value at a specified node with a postprocessor?  Probably not that useful.  Which eigenvector would it output if you could?

File: modules/solid_mechanics/test/tests/modal_analysis/modal.i (hunk 6 above)
Comment (maxnezdyur) [reply modules/solid_mechanics/test/tests/modal_analysis/modal.i:131]: Yes we are solving for lambda = w^2

File: modules/solid_mechanics/test/tests/modal_analysis/modal.i (hunk 7 above)
Comment (maxnezdyur) [reply modules/solid_mechanics/test/tests/modal_analysis/modal.i:140]: As far as I can see this is the only way to switch eigenvectors. Later we should
create an output that can get all the converged eigenvalues and output them at
once for exodus.

File: modules/solid_mechanics/test/tests/modal_analysis/tests [hunk 8]
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  issues = '#27295'
+  [modal]
+    type = Exodiff
+    input = 'modal.i'
+    exodiff = 'modal_out.e'
+    exodiff_opts = '-eigen_relative'
```
Comment (maxnezdyur) [inline modules/solid_mechanics/test/tests/modal_analysis/tests:7]: This is for eigenvectors since -1 * vector is still a correct eigenvectors. Had problems in parallel.

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 9]
```diff
@@ -9,21 +125,21 @@ Other applications of frequency domain dynamics are: (1) computation of band str
...
+where $E$ is the modulus of elasticity, and $\rho$ is the density.
 To convert to the frequency domain, we consider that a plane wave given by
 \begin{equation}
   u(x,t)= B e^{i(\omega t - k x)}
   \label{eq2}
 \end{equation}
-is a solution to [eq1] where $B$ is in general a complex number depending on the boundary conditions, $k^2=\omega^2\rho/E$ is the wave number, $\omega=2 \pi f$ where $f$ is the frequency, and $i=\sqrt{-1}$.  
-By assuming [eq2] is a solution to [eq1] we can solve [eq1] in the frequency domain by taking a Fourier transform of $u(x,t)$ to get $U(x,\omega)$.  The frequency domain version of [eq1] is the Helmholtz equation given by
+is a solution to [eq1] where $B$ is in general a complex number depending on the boundary conditions, $k^2=\omega^2\rho/E$ is the wave number, $\omega=2 \pi f$ where $f$ is the frequency, and $i=\sqrt{-1}$.
+By assuming [eq2] is a a solution to [eq1] we can solve [eq1] in the frequency domain by taking a Fourier transform of $u(x,t)$ to get $U(x,\omega)$.  The frequency domain version of [eq1] is the Helmholtz equation given by
```
Comment (lynnmunday) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:142]:
```suggestion
By assuming [eq2] is a solution to [eq1] we can solve [eq1] in the frequency domain by taking a Fourier transform of $u(x,t)$ to get $U(x,\omega)$.  The frequency domain version of [eq1] is the Helmholtz equation given by
```

File: modules/solid_mechanics/test/tests/modal_analysis/modal.i [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+[]
+[Executioner]
+  type = Eigenvalue
+  solve_type = KRYLOVSCHUR
+  which_eigen_pairs = SMALLEST_MAGNITUDE
+  n_eigen_pairs = 2
+  n_basis_vectors = 5
+  petsc_options = '-eps_monitor_all -eps_view'
+  petsc_options_iname = '-st_type -eps_target -st_pc_type -st_pc_factor_mat_solver_type'
+  petsc_options_value = 'sinvert 0 lu mumps'
```
Comment (lynnmunday) [inline modules/solid_mechanics/test/tests/modal_analysis/modal.i:126]: I've never seen the eigenvalue executioner, how difficult was it to get a solution from this and is it easy to understand convergence?

File: modules/solid_mechanics/test/tests/modal_analysis/tests (hunk 8 above)
Comment (lynnmunday) [reply modules/solid_mechanics/test/tests/modal_analysis/tests:7]: Are the eigenvectors scaled?

**Review by lynnmunday — APPROVED** [review]
General comments but the write up is great.  Thanks!

File: modules/solid_mechanics/test/tests/modal_analysis/modal.i (hunk 10 above)
Comment (maxnezdyur) [reply modules/solid_mechanics/test/tests/modal_analysis/modal.i:126]: `-st_type -eps_target` make is easier to converge. SLEPc has some docs about what to do. The other options was because of problems with exodiff with serial vs parallel.

File: modules/solid_mechanics/test/tests/modal_analysis/tests (hunk 8 above)
Comment (maxnezdyur) [reply modules/solid_mechanics/test/tests/modal_analysis/tests:7]: What do you mean scaled? I meant the eigenvectors make point in the opposite direction depending on parallel vs serial.

**Comment by maxnezdyur** [comment]
@GiudGiud 
Good for review

File: modules/solid_mechanics/test/tests/modal_analysis/tests (hunk 8 above)
Comment (lynnmunday) [reply modules/solid_mechanics/test/tests/modal_analysis/tests:8]: The overall shape is the same but it has an arbitrary scale, like multiplying all of the displacements of these eigenvectors by 5 is still the same eigenvector.  Sometimes the eigenvectors are normalized to have a maximum value of 1, just wondering if they do this here.

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 11]
```diff
@@ -1,5 +1,121 @@
...
+
+In the Kernels block, we define the kernels that contribute to the "A" and "B"
+matrices in the eigenvalue problem. The StressDivergenceTensor kernels represent the stress divergence terms and contribute to the "A"
+matrix. The ADCoefReaction kernels
+represent the mass terms and contribute to the "B" matrix. The
+[!param](/Kernels/CoefReaction/coefficient) in [kernels/CoefReaction.md] is set
+to a negative value which corresponds to a positive density. The
+`extra_vector_tags = 'eigen'` parameter is used to indicate that these kernels
+contribute to "B" matrix. In other literature the A matrix would be the
+stiffness matrix (K) and B would be the Mass matrix(M).
```
Comment (GiudGiud) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:73]:
```suggestion
stiffness matrix (K) and B would be the Mass matrix (M).
```

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 12]
```diff
@@ -1,5 +1,121 @@
...
+[!param](/Executioner/Eigenvalue/n_eigen_pairs) parameter sets the number of eigenvalue pairs to compute.
+
+!listing test/tests/modal_analysis/modal.i block=Executioner id=m_exec
+caption=Eigenvalue Executioner
+
+To output all the eigenvalues solved in the system we can use the
+[vectorpostprocessors/Eigenvalues.md] vectorpostprocessor. While we have solved for the two smallest
+eigenvalues, currently MOOSE only has the ability to output a single eigenvector. To
+change which eigenvector is outputted adjust the index in
+[!param](/Problem/EigenProblem/active_eigen_index)and rerun the simulation.
```
Comment (GiudGiud) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:103]:
```suggestion
[!param](/Problem/EigenProblem/active_eigen_index) and rerun the simulation.
```

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 13]
```diff
@@ -1,5 +1,121 @@
...
+
+!listing test/tests/modal_analysis/modal.i block=VectorPostprocessors Problem
+id=m_prob caption=Eigenvalues
+
+For this coarse mesh example, we were able to determine the first two
+$\omega$'s: 645.1 Hz and 855 Hz. Which are close to the theoretical result of
+509 Hz and 763 Hz, and as the mesh is refined the results would converge on to the
+true solutions. The results are similar to the frequency response function as
```
Comment (GiudGiud) [inline modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:111]: did anyone check?
what refinement do you need?

**Review by GiudGiud — CHANGES_REQUESTED** [review]
looks good, merging after this

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md (hunk 13 above)
Comment (maxnezdyur) [reply modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:111]: With an uniform refinement of 3, we get 513hz and 758hz. IIRC, since our poission's ratio isn't 0 it won't match perfectly. I added this to contrast the original method, so I didn't change any material properties or mesh resolution that was used.

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md [hunk 14]
```diff
@@ -1,5 +1,121 @@
...
+\end{equation}
+This equation represents the generalized eigenvalue problem for modal analysis.
+The terms on the left side, which involve the stress divergence, will contribute
+to the "A" matrix, and the terms on the right side, which are the mass terms, will contribute
+to the "B" matrix.
+
+
+## Cantilever Beam Example
+
+The cantilever beam shown in [cantilever] is subjected to a time harmonic force on the right side in the out-of-plane and vertical directions.  In this example the frequency of the time varying load is swept over a range.  The displacement at the midpoint of the beam is recorded at each frequency.  This type of output, displacement as a function of frequency, is a frequency response function (FRF) or transfer function.  Frequencies corresponding to the displacement peaks in the FRF indicate natural frequencies/modes.
```
[the remark this answers is not in the document]
Comment (maxnezdyur) [reply modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md:41]: Different directions. This is a 3D problem.

**Review by GiudGiud — APPROVED** [review]
